File: app/src/main/java/com/example/pam_project/features/lists/edit/EditListPresenter.java
package com.example.pam_project.features.lists.edit;

import com.example.pam_project.features.lists.list.ListInformation;
import com.example.pam_project.repositories.categories.CategoriesRepository;
import com.example.pam_project.repositories.lists.ListsRepository;
import com.example.pam_project.utils.schedulers.SchedulerProvider;

import java.lang.ref.WeakReference;

import io.reactivex.disposables.Disposable;

public class EditListPresenter {
    private final SchedulerProvider provider;
    private final CategoriesRepository categoriesRepository;
    private final ListsRepository listsRepository;
    private final WeakReference<EditListView> view;
    private Disposable fetchCategoriesDisposable;
    private Disposable updateListDisposable;
    private Disposable deleteListDisposable;

    public EditListPresenter(final SchedulerProvider provider,
                             final CategoriesRepository categoriesRepository,
                             final ListsRepository listsRepository,
                             final EditListView view) {
        this.provider = provider;
        this.categoriesRepository = categoriesRepository;
        this.listsRepository = listsRepository;
        this.view = new WeakReference<>(view);
    }

    public void onViewAttached(final long id) {
        if (view.get() != null) {
            fetchCategories(id);
        }
    }

    private void fetchCategories(final long id) {
        fetchCategoriesDisposable = categoriesRepository.getCategories()
                .subscribeOn(provider.computation())
                .observeOn(provider.ui())
                .subscribe(model -> {
                    if (view.get() != null) {
                        view.get().bindCategories(model);
                        fetchList(id);
                    }
                }, this::onCategoriesReceivedError);
    }

    private void fetchList(final long id) {
        ListInformation model = listsRepository.getList(id);
        if (view.get() != null)
            view.get().bindList(model);
    }

    private void onCategoriesReceivedError(final Throwable throwable) {
        // TODO
    }

    public void updateList(final long id, final String name, final Long categoryId) {
        updateListDisposable = listsRepository.updateList(id, name, categoryId)
                .subscribeOn(provider.computation())
                .observeOn(provider.ui())
                .subscribe(this::onListUpdated, this::onListUpdatedError);
    }

    private void onListUpdated() {
        // TODO
    }

    private void onListUpdatedError(final Throwable throwable) {
        // TODO
    }

    public void deleteList(final long id){
        deleteListDisposable = listsRepository.deleteList(id)
                .subscribeOn(provider.computation())
                .observeOn(provider.ui())
                .subscribe(this::onListDeleted, this::onListDeletedError);
    }

    private void onListDeleted() {
        if (view.get() != null) {
            view.get().onListDelete();
        }
    }

    private void onListDeletedError(final Throwable throwable) {
        // TODO
    }

    public void onViewDetached() {
        if(fetchCategoriesDisposable != null)
            fetchCategoriesDisposable.dispose();
        if(updateListDisposable != null)
            updateListDisposable.dispose();
        if(deleteListDisposable != null)
            deleteListDisposable.dispose();
    }

}

File: app/src/main/java/com/example/pam_project/features/lists/edit/EditListPresenter.kt
package com.example.pam_project.features.lists.edit

import com.example.pam_project.features.categories.list.CategoryInformation
import com.example.pam_project.features.lists.list.ListInformation
import com.example.pam_project.repositories.categories.CategoriesRepository
import com.example.pam_project.repositories.lists.ListsRepository
import com.example.pam_project.utils.schedulers.SchedulerProvider
import io.reactivex.disposables.Disposable
import java.lang.ref.WeakReference

class EditListPresenter(private val provider: SchedulerProvider,
                        private val categoriesRepository: CategoriesRepository,
                        private val listsRepository: ListsRepository,
                        view: EditListView) {
    private val view: WeakReference<EditListView?> = WeakReference(view)
    private var fetchListDisposable: Disposable? = null
    private var fetchCategoriesDisposable: Disposable? = null
    private var updateListDisposable: Disposable? = null
    private var deleteListDisposable: Disposable? = null
    fun onViewAttached(id: Long) {
        if (view.get() != null) {
            fetchCategories(id)
        }
    }

    private fun fetchCategories(id: Long) {
        fetchCategoriesDisposable = categoriesRepository.categories
                .subscribeOn(provider.computation())
                .observeOn(provider.ui())
                .subscribe({ model: List<CategoryInformation?>? ->
                    if (view.get() != null) {
                        view.get()!!.bindCategories(model)
                        fetchList(id)
                    }
                }) { throwable: Throwable -> onCategoriesReceivedError(throwable) }
    }

    private fun fetchList(id: Long) {
        fetchListDisposable = listsRepository.getList(id)
                .subscribeOn(provider.computation())
                .observeOn(provider.ui())
                .subscribe({ model: ListInformation? -> onListReceived(model) }) { throwable: Throwable -> onListReceivedError(throwable) }
    }

    private fun onListReceived(model: ListInformation?) {
        if (view.get() != null) view.get()!!.bindList(model)
    }

    private fun onListReceivedError(throwable: Throwable) {
        if (view.get() != null) {
            view.get()!!.onListReceivedError()
        }
    }

    private fun onCategoriesReceivedError(throwable: Throwable) {
        if (view.get() != null) {
            view.get()!!.onCategoriesReceivedError()
        }
    }

    fun updateList(id: Long, name: String, categoryId: Long?) {
        updateListDisposable = listsRepository.updateList(id, name, categoryId!!)
                .subscribeOn(provider.computation())
                .observeOn(provider.ui())
                .subscribe({}) { throwable: Throwable -> onListUpdatedError(throwable) }
    }

    private fun onListUpdatedError(throwable: Throwable) {
        if (view.get() != null) {
            view.get()!!.onListUpdatedError()
        }
    }

    fun deleteList(id: Long) {
        deleteListDisposable = listsRepository.deleteList(id)
                .subscribeOn(provider.computation())
                .observeOn(provider.ui())
                .subscribe({ onListDeleted() }) { throwable: Throwable -> onListDeletedError(throwable) }
    }

    private fun onListDeleted() {
        if (view.get() != null) {
            view.get()!!.onListDelete()
        }
    }

    private fun onListDeletedError(throwable: Throwable) {
        if (view.get() != null) {
            view.get()!!.onListDeletedError()
        }
    }

    fun onDeletePressed() {
        if (view.get() != null) {
            view.get()!!.showDeleteDialog()
        }
    }

    fun onViewDetached() {
        if (fetchListDisposable != null) fetchListDisposable!!.dispose()
        if (fetchCategoriesDisposable != null) fetchCategoriesDisposable!!.dispose()
        if (updateListDisposable != null) updateListDisposable!!.dispose()
        if (deleteListDisposable != null) deleteListDisposable!!.dispose()
    }

}